devito/types/grid.py

```diff
@@ -421,3 +425,125 @@ class Interior(SubDomain):
+
+    Set the bounds of the subdomains. The required format is:
+    (xm, xM, ym, yM, ...) where xm is a vector specifying
+    the number of grid points inwards from the 'left' boundary in the
+    first grid dimension that each subdomain starts. xM is a vector
+    specifying the number of grid points inwards from the 'right' of
+    the domain in the first grid dimension that each subdomain ends.
+    ym and yM are the equivalents for the second grid dimension.
+
+    >>> xm = np.array([1, Nx/2+1], dtype=np.int32)
+    >>> xM = np.array([Nx/2+1, 1], dtype=np.int32)
+
+    Along a dimension where all bounds are the same we can use the
+    following shorthand:
+
+    >>> ym = 1 # which is equivalent to 'np.array([1, 1], dtype=np.int32)'
+    >>> yM = 1
+
+    Combine the data into the required form:
+
+    >>> bounds = (xm, xM, ym, yM)
+
+    Create our set of subdomains passing the number of domains and the
+    bounds:
+
+    >>> my_sd = MySubdomains(N=n_domains, bounds=bounds)
+
+    Create a grid and iterate a function within the define subdomains:
+
+    >>> grid = Grid(extent=(Nx, Ny), shape=(Nx, Ny), subdomains=(my_sd, ))
+    >>> f = Function(name='f', grid=grid, dtype=np.int32)
+    >>> eq = Eq(f, f+1, subdomain=grid.subdomains['mydomains'])
+    >>> op = Operator(eq)
+    >>> summary = op.apply()
+    >>> f.data
+    Data([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+          [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+          [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+          [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+          [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+          [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+          [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+          [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+          [0, 1, 1, 1, 1, 1, 1, 1, 1, 0],
+          [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=int32)
+    """
+
+    def __init__(self, **kwargs):
+        super(SubDomainSet, self).__init__()
+        self._n_domains = kwargs.get('N', 1)
+        self._bounds = kwargs.get('bounds', None)
+
+    def __subdomain_finalize__(self, dimensions, shape):
+        super(SubDomainSet, self).__subdomain_finalize__(dimensions, shape)
+        n = Dimension(name='n')
+        self._implicit_dimension = n
+
+    @property
+    def n_domains(self):
+        return self._n_domains
+
+    @property
+    def bounds(self):
+        return self._bounds
+
+    @cached_property
+    def _implicit_eq_dat(self):
+        if not len(self._bounds) == 2*len(self.dimensions):
+            raise ValueError("Left and right bounds must be supplied for each dimension")
+        n_domains = self.n_domains
+        i_dim = self._implicit_dimension
+        b_f = {}
+        dat = []
+        # Organise the data contained in 'bounds' into a form such that the
+        # associated implicit equations can easily be created.
+        for j in range(0, len(self._bounds)):
+            index = floor(j/2)
+            d = self.dimensions[index]
+            if j % 2 == 0:
```

Comment:
I have a feeling this code may benefit from rewriting the loop as:

```
for i, fname in enumerate(self._bounds):
     ...
```

because then you can drop `index = ...`, while you can get the `fname` by adding a private property `fnames` (feel free to pick a better name) returning `{self._bounds[0]: d.min_name, self._bounds[1]: d.max_name}`, which ultimately will allow you to drop the `if ...` , while making it easier to understand what we are trying to achieve

Reply:
I'm not quite sure what you mean by the above. Where are you suggesting that `fnames` is coming from?

Reply:
sorry, I thought bounds were provided as a tuple of 2-tuples, ie, ``((bounds_xm, bounds_xM), (bounds_ym, bounds_yM))``, then it would have been easier to have a property ``fnames`` returning a dict which associates a name to each of that bounds, but I see now that actually bounds is a flatten tuple, so never mind, no worries !